### ISSUE TITLE
bug: Thunderbird creates new local folders account on every launch

### ISSUE THREAD

**Issue by nfelber** [nix-community/home-manager#5031]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

## Issue

On every launch, Thunderbird adds a new line `user_pref("mail.account.account<N>.server", "server1");` to `prefs.js` of the profile being used, with `<N>` a value incrementing in steps of 1. The parameter `mail.account.lastKey` is also incremented to be equal to the largest `<N>`. It doesn't affect the accounts declared in the home-manager nix config.

## Example

`prefs.js` after launching and closing Thunderbird 5 times:

```js
// ...
// Lines weirdly added on each launch
user_pref("mail.account.account1.server", "server1");
user_pref("mail.account.account2.server", "server1");
user_pref("mail.account.account3.server", "server1");
user_pref("mail.account.account4.server", "server1");
user_pref("mail.account.account5.server", "server1");
// Accounts normally declared in the home-manager nix config
user_pref("mail.account.account_576ba7c2e4abb7184ca409154dbbbd5306c1a80747fbce4148ea6271fd21e776.identities", "id_576ba7c2e4abb7184ca409154dbbbd5306c1a80747fbce4148ea6271fd21e776");
user_pref("mail.account.account_576ba7c2e4abb7184ca409154dbbbd5306c1a80747fbce4148ea6271fd21e776.server", "server_576ba7c2e4abb7184ca409154dbbbd5306c1a80747fbce4148ea6271fd21e776");
user_pref("mail.account.account_6f0a53fb760ead798eff28586351f9cda5211f8d18df529aaca8391c814ba87a.identities", "id_6f0a53fb760ead798eff28586351f9cda5211f8d18df529aaca8391c814ba87a");
user_pref("mail.account.account_6f0a53fb760ead798eff28586351f9cda5211f8d18df529aaca8391c814ba87a.server", "server_6f0a53fb760ead798eff28586351f9cda5211f8d18df529aaca8391c814ba87a");
user_pref("mail.account.account_8758902c10feb2ee51fb2a4a0afcde74fd6b11057e0be55d98d4d85335e8deee.identities", "id_8758902c10feb2ee51fb2a4a0afcde74fd6b11057e0be55d98d4d85335e8deee");
user_pref("mail.account.account_8758902c10feb2ee51fb2a4a0afcde74fd6b11057e0be55d98d4d85335e8deee.server", "server_8758902c10feb2ee51fb2a4a0afcde74fd6b11057e0be55d98d4d85335e8deee");
// lastKey equals largest account number
user_pref("mail.account.lastKey", 5);
// ...
```

## Possible explanation

The `thunderbird.nix` module sets the `mail.accountmanager.accounts` parameter in the `user.js` file of the profile as the comma separated concatenation of the names of the accounts declared in the home-manager nix config (these names are based on hashes of the actual account names). This parameter controls which accounts show up in Thunderbird's folder pane and in which order[^1]. On startup, Thunderbird reads the `user.js` file and copies all its parameters to `prefs.js` (overwriting existing ones).

In my understanding: as no local folders account is listed in `mail.accountmanager.accounts`, Thunderbird creates a new one (with an incrementally increasing number in the name to avoid conflicts with existing ones) and appends it to `mail.accountmanager.accounts` in `prefs.js` to make the local folders appear in the folder pane. On subsequent launches, the value of `mail.accountmanager.accounts` in `prefs.js` will be overwritten by the value from `user.js`, resulting in a cycle.

## Workaround

I was able to solve both the automatic local folders account creation and the ordering issues by manually setting the `mail.accountmanager.accounts` value (with desired ordering and `account1` appended) using the `settings` option of the `thunderbird.nix` module.

```nix
{
  programs.thunderbird = {
    enable = true;
    profiles = {
      nf = {
        isDefault = true;
        settings = let
          toHashedName = x: "account_" + (builtins.hashString "sha256" x);
          # List accounts in the order they will appear in the folder pane
          accounts = [
            "pascaline"
            "gmail"
            "epfl"
          ];
        in {
          "mail.accountmanager.accounts" = lib.concatStringsSep "," ((map toHashedName accounts) ++ ["account1"]); 
        };
      };
    };
  };
}
```

[^1]: A related issue is that there is currently no way to declare the order in which the accounts should appear using the `thunderbird.nix` module, and any modification to this order made in the UI (i.e. written to `prefs.js`) will be overwritten (by `user.js`) on the next launch.

### Maintainer CC

`meta.maintainers`: @d-dervishi @jkarlson

Thanks!

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 6.1.77, NixOS, 24.05 (Uakari), 24.05.20240211.f9d39fb`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.11"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by kira-bruneau** [comment]
I was running into the same problem, and realized it was because I wasn't defining any IMAP configuration for my accounts. It looks like if IMAP isn't configured, thunderbird will automatically remove it from the list of accounts, causing this issue.

Maybe the module should assert for this? Or default to an empty imap configuration if one isn't given: https://superuser.com/questions/1764409/how-to-send-email-with-thunderbird-without-configuring-an-incoming-email-service

**Comment by alois31** [comment]
Not stale. @nfelber would you mind submitting the fix as a PR?

**Comment by nfelber** [comment]
No problem, I'm a bit busy right now but I'll gladly look into it after Christmas.

**Comment by jonocodes** [comment]
@nfelber do you want to pass on the PR so someone else can merge perhaps?
Thanks for your work.

**Comment by nfelber** [comment]
Yes, I'll be happy to pass on the PR if anyone is interested since I'm struggling to find any time to work on it at the moment. How can I do that?

**Comment by jonocodes** [comment]
Perhaps you can generate a patch/diff file of your work and post it here.

**Comment by nfelber** [comment]
Actually I will merge it myself, I forgot the PR was pretty small. Just needs some additional review and approval.

**Comment by jonocodes** [comment]
I believe this is still a problem. Should it be fixed by https://github.com/nix-community/home-manager/pull/6310/ ?

**Comment by nfelber** [comment]
> Should it be fixed by [#6310](https://github.com/nix-community/home-manager/pull/6310) ?

Yes, it should—*in theory*—be fixed by [#6310](https://github.com/nix-community/home-manager/pull/6310). At least I don't experience the issue anymore. Can you provide more details?

**Comment by jonocodes** [comment]
> > Should it be fixed by [#6310](https://github.com/nix-community/home-manager/pull/6310) ?
> 
> Yes, it should—_in theory_—be fixed by [#6310](https://github.com/nix-community/home-manager/pull/6310). At least I don't experience the issue anymore. Can you provide more details?

Sorry, I dont boot the machine that uses this much so it takes time to reply...

In /home/jono/.thunderbird/default

> grep -i lastkey prefs.js
user_pref("mail.account.lastKey", 14);

Then I start thunderbird and then look again

> grep -i lastkey prefs.js
user_pref("mail.account.lastKey", 15);

Since it keeps making copies of definitions of my folders, it is not saving my preferences for them.

> nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-linux"`
 - host os: `Linux 6.14.8-asahi, NixOS, 25.05 (Warbler), 25.05.20250823.b1b3291`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.4`
 - channels(root): `"nixos-24.11"`
 - nixpkgs: `/nix/store/x2k3b38vdbrpxsf262lpwlmpf8nggbxj-source`
